[HttpKernel] Correctly merging cache directives in HttpCache/ResponseCacheStrategy

aschempp opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #26245, #26352, #28872
| License       | MIT
| Doc PR        | -

This PR is a first draft to fix the incorrect merging of private and other cache-related headers that are not meant for the shared cache but the browser (see mentioned issues).

The existing implementation of `HttpFoundation\Response` is very much tailored to the `HttpCache`, for example `isCacheable` returns `false` if the response is `private`, which is not true for a browser cache. That is why my implementation does not longer use much of the response methods. They are however still used by the `HttpCache` and we should keep them as-is. FYI, the `ResponseCacheStrategy` does **not** affect the stored data of `HttpCache` but is only applied to the result of multiple merged subrequests/ESI responses.

I did read up a lot on RFC2616 as a reference. [Section 13.4](https://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.4) gives an overall view of when a response MAY be cached. [Section 14.9.1](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9.1) has more insight into the `Cache-Control` directives. 

Here's a summary of the relevant information I applied to the implementation:

 - > Unless specifically constrained by a cache-control (section 14.9) directive, a caching system MAY always store a successful response (see section 13.8) as a cache entry, MAY return it without validation if it is fresh, and MAY return it after successful validation.

    A response without cache control headers is totally fine, and it's up to the cache (shared or private) to decide what to do with it. That is why the implementation does not longer set `no-cache` if no `Cache-Control` headers are present.

 - > A response received with a status code of 200, 203, 206, 300, 301 or 410 MAY be stored […] unless a cache-control directive prohibits caching.

    > A response received with any other status code (e.g. status codes 302 and 307) MUST NOT be returned […] unless there are cache-control directives or another header(s) that explicitly allow it.

    This is what `ResponseCacheStrategy::isUncacheable` implements to decide whether a response is not cacheable at all. It differs from `Response::isCacheable` which only returns true if there are actual `Cache-Control` headers.

 - > [Section 13.2.3](https://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.2.3): When a response is generated from a cache entry, the cache MUST include a single Age header field in the response with a value equal to the cache entry's current_age.

    That's why the implementation **always** adds the `Age` header. It takes the oldest age of any of the responses as common denominator for the content.

 - > [Section 14.9.3](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9.3): If a response includes an s-maxage directive, then for a shared cache (but not for a private cache), the maximum age specified by this directive overrides the maximum age specified by either the max-age directive or the Expires header.

    This effectively means that `max-age`, `s-maxage` and `Expires` must all be kept on the response. My implementation assumes that we can only do that if they exist in **all** of the responses, and then takes the lowest value of any of them. Be aware the implementation might look confusing at first. Due to the fact that the `Age` header might come from another subresponse than the lowest expiration value, the values are stored relative to the current response date and then re-calculated based on the age header.


The Symfony implementation did not and still does not implement the full RFC. As an example, some of the `Cache-Control` headers (like `private` and `no-cache`) MAY actually have a string value, but the implementation only supports boolean. Also, [Custom `Cache-Control` headers](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9.6) are currently not merged into the final response.


**ToDo/Questions:**

 1. [Section 13.5.2](https://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.5.2) specifies that we must add a [`Warning 214 Transformation applied`](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.46) if we modify the response headers.

 2. Should we add an `Expires` headers based on `max-age` if none is explicitly set in the responses? This would essentially provide the same information as `max-age` but with support for HTTP/1.0 proxies/clients.

 3. I'm not sure about the implemented handling of the `private` directive. The directive is currently only added to the final response if it is present in all of the subresponses. This can effectively result in no cache-control directive, which does not tell a shared cache that the response must not be cached. However, adding a `private` might also tell a browser to actually cache it, even though non of the other responses asked for that.

 4. > [Section 14.9.2](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9.2): The purpose of the `no-store` directive is to prevent the inadvertent release or retention of sensitive information […]. The `no-store` directive applies to the entire message, and MAY be sent either in a response or in a request. If sent in a request, a cache MUST NOT store any part of either this request or any response to it. If sent in a response, a cache MUST NOT store any part of either this response or the request that elicited it.

    I have not (yet) validated whether the `HttpCache` implementation respects any of this.

 5. As far as I understand, the current implementation of [`ResponseHeaderBag::computeCacheControlValue`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php#L313) is incorrect. `no-cache` means a response [must not be cached by a shared or private cache](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9.1), which overrides `private` automatically.

 5. The unit tests are still very limited and I want to add plenty more to test and sort-of describe the implementation or assumptions on the RFC.

/cc @nicolas-grekas 

#SymfonyConHackday2018

aschempp committed:
[HttpKernel] Correctly merging cache directives in HttpCache/ResponseCacheStrategy

nicolas-grekas commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -28,30 +28,70 @@
+        $this->storeRelativeAgeDirective('s-maxage', $response->headers->getCacheControlDirective('s-maxage'), $age);
+
+        $expires = $response->getExpires();
+        $expires = null === $expires ? null : $expires->format('U') - $response->getDate()->format('U');
```
I suggest to swap the two expressions of the ternary op, thus checl `null !== $expire` to help reading with precedence

nicolas-grekas commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -28,30 +28,70 @@
+
+        $expires = $response->getExpires();
+        $expires = null === $expires ? null : $expires->format('U') - $response->getDate()->format('U');
+        $this->storeRelativeAgeDirective('expires', $expires < 0 ? null : $expires, 0);
```
same here: `0 < $expires ? $expires : null`

nicolas-grekas commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -64,33 +104,97 @@ public function update(Response $response)
+     *
+     * @see https://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.4
+     *
+     * @param Response $response
```
can be removed

nicolas-grekas commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -64,33 +104,97 @@ public function update(Response $response)
+     *
+     * @param Response $response
+     *
+     * @return bool
```
can be put as return type

aschempp replied:
This is for Symfony 3.4, so PHP 5.5

nicolas-grekas commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -64,33 +104,97 @@ public function update(Response $response)
+     *
+     * @param string   $cacheKey
+     * @param int|null $value
+     * @param int      $age
```
should be replaced by real type ints

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Tests/HttpCache/ResponseCacheStrategyTest.php:
```diff
@@ -88,8 +88,8 @@ public function testMasterResponseNotCacheableWhenEmbeddedResponseRequiresValida
-        $this->assertTrue($masterResponse->headers->hasCacheControlDirective('no-cache'));
-        $this->assertTrue($masterResponse->headers->hasCacheControlDirective('must-revalidate'));
+//        $this->assertTrue($masterResponse->headers->hasCacheControlDirective('no-cache'));
+//        $this->assertTrue($masterResponse->headers->hasCacheControlDirective('must-revalidate'));
```
cannot stay as is (same below)

nicolas-grekas reviewed (changes requested):
Some minor comments to start the review.

My most important comment is that this should be merged as a new feature to me, thus rebased against master.

fabpot commented:
If you have time @mpdude, I would love to get your review on this one.

mpdude commented:
I'd really like to help here! It's a bit difficult, though, as the description does not directly make clear what we're after. After reading the linked issues my understanding is that we want to enable the `ResponseCacheStrategy` to deal with `private` responses and come up with a better result than `private, no-cache, must-revalidate`.

Before discussing the implementation, can we agree on what the resulting behavior needs to be?

I have reviewed the existing tests and think that the merge rules implemented in `ResponseCacheStrategy` would need to be as follows.

0. If we have no embedded response, leave the master response alone.
1. If one of the responses is `private`, the result must be `private`. Otherwise, all responses are `public`.
2. Validation-type headers (`Last-Modified`, `ETag`) need to be removed. Validation cannot reasonably be used on merged responses. 
3. If not all responses provide expiration times, we cannot compute an expiration time for the result and need to stick with `no-cache, must-revalidate`.
4. If the result is `public`, compute the final `s-maxage` by looking for the smallest value among all responses, considering `s-maxage`, `max-age` and `Expires` in this order.
5. When all responses provide either `max-age` or `Expires`, compute the final `max-age` by looking for the smallest value among all responses, considering `max-age` and `Expires` in this order.
6. Calculate `Age` in a way that the resulting response becomes stale as soon as the lowest TTL of any response expires.

Does that make sense?

aschempp commented:
0. agreed

1. If not all of the responses are marked as `public`, the final should not have `public` imho. Not having a `public` directive is not the same as not having any directive. In the same sense, having one `private` response (allow browser caching) does not mean all responses can be cached by the browser. Only if all other are either `public` or `private` a browser should be allowed to cache the result. *FYI: this might not be implemented in my PR*.

2. agreed

3. Why do we merge multiple header values that are not meant to be the same?
    - `s-maxage` is for shared caches only
    - `max-age` is for private caches as well as for shared caches if no `s-maxage` is defined
    - `Expires` headers are similar to `max-age`, but mostly useful for HTTP/1.0 protocol. `Expires` can existing in combination with `max-age` to control or prohibit caching for HTTP/1.0 servers, especially if they do not support features like `must-revalidate`.

4. same as 3.

5. A `private` directive without `max-age` is perfectly fine. Each cache definition is valid without other. It's up to the cache server to decide on a cache duration if none is given. This assumes **all** responses are marked as private but all or some of them do not have a `max-age`.

6. The age has nothing to do with expiration time. `Age` is determined by the difference between `Date` header (when the response has been added to cache) and *now*. After merging multiple responses, age must represent the oldest response's age imho.


I still don't understand why Symfony generally adds a `no-cache, must-revalidation`. Quoting from #26245 

> >A `no-cache` directive essentially tells all caches (and that includes the Browser, which is a private cache) to no cache the response.

> I think you're misreading the RFC you quoted? `no-cache` means "don't serve this from a cache without revalidation", `private` means "store in private caches only". `no-store` is the way of stating "don't keep this in caches".

You're somewhat right with my incorrect interpretation. I wasn't very sure about revalidation, so I looked it up: https://stackoverflow.com/questions/18148884/difference-between-no-cache-and-must-revalidate#19938619
So essentially, `no-cache, must-revalidate` allows a cache to store it but never to use it without revalidation. And we are removing both possible revalidation options (Etag and Last-Modified, see 2.), which essentially results in a useless waste of storage? Also, `no-cache` implies `must-revalidate`, so the second directive is not useful either?

Toflar commented:
Any progress here? The current implementation breaks client cache headers when working with ESI fragments :(

nicolas-grekas commented:
friendly ping @aschempp

aschempp commented:
I have updated the PR to fix some mentioned coding style and implemented point 1 in [my comment](https://github.com/symfony/symfony/pull/26532#issuecomment-375952282).

However, most of the notes in my thoughts in the last comment still stand, so I guess we should also ping @mpdude 😁 

Before we discuss the fine details of this though, we should agree or even write down our thoughts on the supposed *default behavior*. Symfony [**always** adds cache control headers](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php#L310) which I consider strange. Especially the second if-condition means I can have an `Expires` header but the `ResponseHeaderBag` will add a cache-control directive that "breaks" this for any HTTP/1.1 cache.

According the the RFC, not having a `Cache-Control` header is perfectly fine, but that is not possible with Symfony. Is this behavior considered correct @mpdude / @nicolas-grekas ? And if we want to be conservative (prevent caching if no caching headers are given), I think we should still correctly handle HTTP/1.0 headers?

aschempp commented:
Is there anything I can do to push this forward?

aschempp commented:
Rebase is done. The change from https://github.com/aschempp/symfony/commit/3d27b5946db147da7d96b14bb2a89b8a25d8a58b#diff-e5a339b48cec0fa0a4d0c5c4c705f568R75 has been dropped, but I preserved the unit test and they are still passing.

aschempp commented:
I just looked at the disabled unit tests and will follow up with a few changes shortly.

aschempp commented:
Sorry for taking a while. I tried to implement merging of `Last-Modified` and `Etag` but realized this won't work reliably. They now render a response uncachable without additional expiry headers.

---

One thing I'm not sure about is the default Response behavior. If there are not `Cache-Control` headers, Symfony always adds `private, must-revalidate`. https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php#L309-L331

First of all, `private` means a browser is allowed to cache it, which is not the same as _no Cache-Control provided_, ie. probably not cacheable. Secondly, `must-revalidate` does not make sense because there is never any validation information (`Last-Modified` and `Etag` are removed), a validation will never be possible.

aschempp commented:
I have removed two assertions in the HttpCacheTest to fix tests. I'm very well aware that removing assertions to fix a test is not really a practical solution. However, as explained in https://github.com/symfony/symfony/issues/26245#issuecomment-375630190, I think the assertions (and Symfony behavior) were wrong in the first place …

ausi commented on src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php:
```diff
@@ -1242,7 +1242,6 @@ public function testEsiCacheForceValidation()
         $this->assertEquals('Hello World! My name is Bobby.', $this->response->getContent());
         $this->assertNull($this->response->getTtl());
         $this->assertTrue($this->response->mustRevalidate());
-        $this->assertTrue($this->response->headers->hasCacheControlDirective('private'));
```
Shouldn’t this be changed to `$this->assertFalse(…)` instead of removing the assertion?

aschempp replied:
I don't think so, as this _tests that esi cache forces validation_, and that case is validated with the line above. The cache header is not related to esi cache validation.

mpdude replied:
I guess the `Hello World` response is `private` by default, since is does not explicitly set `public` or a `s-maxage`? 

If so, shouldn’t `private` then be preserved?

mpdude replied:
Or am I confusing behavior of the `Response` class with how this test works?

aschempp commented:
Any follow-up questions for this PR I need to address?

nicolas-grekas commented:
(could you please rebase on latest 3.4?)

mpdude commented:
For me, this PR is too confusing to give meaningful feedback. 

Can we somehow simplify this or break it into smaller pieces we could discuss independently?

aschempp commented:
@mpdude did you look at the diff or at the new version of the file directly? What can I do to simplify the changes? I think they are all related, there are no "individual features" that could be split into smaller pieces.

@nicolas-grekas I will rebase as soon as any questions are resolved. Previously all changes in the old file were irrelevant because it is basically a complete rewrite of the class.

aschempp commented:
Is there anything I can do to get this merged at some point?

fritzmg commented:
Due to the [already mentioned security issue](https://github.com/symfony/symfony/issues/28872#issuecomment-429772627) this should be done as soon as possible …

fritzmg commented:
Due to the [already mentioned security issue](https://github.com/symfony/symfony/issues/28872#issuecomment-429772627) this should be merged as soon as possible.

fritzmg commented:
Due to the [already mentioned security issue](https://github.com/symfony/symfony/issues/28872#issuecomment-429772627) this should be merged as soon as possible.

mpdude commented:
In those places where you changed existing tests, could you leave a comment explaining why the change is necessary and/or correct?

aschempp commented:
I have added unit tests in #29522 to show the incorrect behavior of the `ResponseCacheStrategy`. Also reverted any other changes to unit tests in this PR for now, so we can see what's failing (because the tests are incorrect).

fabpot commented:
@aschempp Can you rebase this pull request? We cannot merge a pull request with a merge commit. Thank you.

stof commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -64,33 +99,116 @@ public function update(Response $response)
-        } elseif (null !== $maxAge = min($this->maxAges)) {
-            $response->setSharedMaxAge($maxAge);
-            $response->headers->set('Age', $maxAge - min($this->ttls));
+        // Last-Modified and Etag headers cannot be merged, they render the response uncacheable
```
This comment does not seem to correspond to the following code, as the following deals with responses *not* having Last-Modified and Etag

dbu replied:
maybe formulate it as "If an embedded response uses Last-Modified or an Etag, the combined response is not cacheable."

hm, but the code is saying that if we have a success status AND no last modified / etag, it certainly *is* cacheable.
should we instead say that if there is either etag or last-modified, we return true?

and to simplify reasoning in this method, i suggest flipping this method over to `keepsFinalResponseCacheable` and switch true / false. and invert the bool where we call the method.

aschempp replied:
First, the method name was a lengthy discussion and finally was @nicolas-grekas's suggestion.

This step determines that according to RFC, a response is **always** cacheable if it has one of the given response codes. Not having any cache-control information does **not** make a response uncacheable, it just does not tell a cache what to do.

dbu replied:
ok with the method name. i prefer "positive" naming over negations, but the field is also done that way round so it could add confusion.

to make this robust and easier to understand, how about saying that as soon as there is etag or last-modified, we return true. and move the status check code to the very end. instead of return true, `return !\in_array($response->getStatusCode(), array(...));` . that would be more explicit i think.

aschempp replied:
That would not be the same!
 - Your suggestion means that a response with either `ETag` or `Last-Modified` *willMakeFinalResponseUncacheable*. That is not what I (tried to) implement
 - The final response is uncache if 
    1. ( it is not a given status code **OR** it has `ETag` or `Last-Modified` )
    2. **AND** it does not have any other caching info (like `max-age`).

The implementation just works the opposite way. 
  1. If it has a given status code and none of the headers, we already know it will not 
      *…MakeFinalResponseUncacheable*
  2. If either it has a different status code **OR** it has one of the headers, we must also check for Cache-Control headers

dbu replied:
right, this is not the same. hm, so if status is 200 and we have no etag/last-modified, we say its ok to cache, even if max-age is set to 0? ah, but then we would still mark the final result with max-age: 0, and same goes for private.

okay, then i agree this is correct, though somewhat counterintuitive. can you mention this in the phpdoc, that cache-control instructions are handled elsewhere?

also, if the fragment has no cache-control header but the master response has max-age: 1 day, would we end up with caching the combined response for a whole day? or should we assume a default max-age when caching something with status 200 and no cache-control instruction? varnish takes 2 minutes in that case, by default...

aschempp replied:
> hm, so if status is 200 and we have no etag/last-modified, we say its ok to cache, even if max-age is set to 0?

No, we don't say its ok to cache. We just determine that this response does not make the final response uncacheable. The final response can still have no useful caching info and therefore be not cacheable.

> if the fragment has no cache-control header but the master response has max-age: 1 day, would we end up with caching the combined response for a whole day?

There is **no difference** between ESI fragments and the master response. They are all sent to the `add` method, and the result of ALL responses (regardless of their type) is added in the `update` method. So if any - regardless if master or fragment - has no mergeable data, nothing will be added to the final response.

dbu replied:
ah right, this happens here: https://github.com/symfony/symfony/pull/26532/files#diff-e5a339b48cec0fa0a4d0c5c4c705f568R212 - if one response has no max-age we set that info to `false`.

so this should indeed be fine. maybe put part of this explanation into the phpdoc?

> We just determine that this response does not make the final response uncacheable. The final response can still have no useful caching info and therefore be not cacheable.

stof commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -64,33 +99,116 @@ public function update(Response $response)
+        // by default (except if the response also has max-age etc.).
+        if (\in_array($response->getStatusCode(), array(200, 203, 300, 301, 410))
+            && null === $response->getLastModified()
+            && null === $response->getEtag()
```
What is missing here is that *having* a Etag or LastModified on an embedded response must make the final response uncacheable if it does not have a max-age or expire (as the validation cannot be done).

aschempp replied:
I agree it's a bit confusing, but thats is what the code does.

The method returns `true` if the final response is uncacheable. So
 1. If the status code is cacheable by default, **and** we do not have a `Last-Modified` or `E-Tag`, return `false` because the response is cacheable
 2. If the status code was invalid or one of the headers was present, only return `false` (= cacheable) if one of the cache headers (`max-age` etc.) is present.
 3. Finally return `true` (= uncacheable) if none of the conditions apply.

dbu replied:
even if there are max-age/expire, the etag / last-modified can still not be validated. 

or does http cache on if-none-match/if-modified-since requests check all embedded ESI too? i think not, as in the main method, we remove etag and last-modified from the master response we send. i think this can be simplified to "if there is any etag / last-modified then the master response is not cacheable"

aschempp replied:
The cache cannot validate ESI responses. But removing ETag or Last-Modified only prevents the response from being validated, it can still be cached for the `max-age` etc. duration. Validation only kicks in **after** `max-age` etc. has expired.

dbu replied:
true that, agree. see my suggestion above for how to restructure the logic to make it easier to understand.

dbu commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -64,33 +99,116 @@ public function update(Response $response)
+            $sMaxage = $this->ageDirectives['s-maxage'] + $this->age;
 
-            return;
+            if ($maxAge !== $sMaxage) {
```
that seems unexpected to me. while it is somewhat redundant, i don't see symfony doing the same in a normal response, so i'd not eliminate the duplication.

aschempp replied:
The duplication is produced by the `ResponseCacheStrategy` itself, because we need to merge `max-age` into `s-maxage` if only the former is set, because another response might only have the later and then `s-maxage` could still be sent.

dbu replied:
we could try to track things separately. but i think in the end it would be good enough to check if the master response has s-maxage set and in that case re-add it regardless?

oh, i just notice that you do not remove s-maxage from the master response, only overwrite it if its different from max-age. could that be a problem when both are 0?

aschempp replied:
All existing Cache-Control headers on the master response are overwritten by https://github.com/symfony/symfony/pull/26532/files/77ae8bc10240721c44a78e3338d5c28dd7c11c55#diff-e5a339b48cec0fa0a4d0c5c4c705f568R130. `s-maxage` is only added if it differs from `max-age`. Be aware that the master response is no difference than any ESI response, it is first passed to `update` so it's Cache-Control values are in the ResponseCacheStrategy already.

dbu replied:
oh, indeed, missed that. then could we track if the master response has an s-maxage set, and if so keep setting it regardless of whether it is different from max-age, to avoid a surprise?

aschempp replied:
We can't set a `s-maxage` of the master response, if the ESI responses do not have that. We can only set a Cache-Control header if **all** responses have a value, and if we can merge them.

dbu replied:
right. there is still the edge case when master response and all fragments have s-maxage == max-age. then we would not set the (redundant) s-maxage because its the same as max-age. this might be a bit confusing, but if "phantom" s-maxage appears its also confusing, so i guess we can ignore the issue.

dbu replied:
genau. i agreed on leaving it as is, even if there is an edge case that could be a bit confusing, where you set both values on all fragments and the master response, but end up having only max-age.

dbu commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -64,33 +99,116 @@ public function update(Response $response)
+        // RFC2616: A response received with any other status code (e.g. status codes 302 and 307)
+        // MUST NOT be returned in a reply to a subsequent request unless there are
+        // cache-control directives or another header(s) that explicitly allow it.
+        $cacheControl = array('max-age', 's-maxage', 'must-revalidate', 'proxy-revalidate', 'public', 'private');
```
does must-revalidate indeed work as intended? that is, does the HttpCache on such a request also check all embedded ESI requests?

aschempp replied:
It does not, and it cannot, at least sure not for ETag, but Last-Modified would be **very** complex as well. This list is taken from the RFC, not from "what might be useful in our case". Symfony always [adds "must-revalidate"](https://github.com/symfony/symfony/blob/77ae8bc10240721c44a78e3338d5c28dd7c11c55/src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php#L116-L118) [by default](https://github.com/symfony/symfony/blob/77ae8bc10240721c44a78e3338d5c28dd7c11c55/src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php#L318), even though that can never be used because of missing validation headers. But @stof and I agreed to not change the existing behavior.

dbu replied:
i read up what must-revalidate is, and its about the handling of stale cache, so when max-age is exceeded. i somehow thought it forces to always use etag/last-modified, but that would be achieved by setting max-age 0 and an etag. sorry for the confusion here.

dbu commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -5,10 +5,6 @@
  *
- * This code is partially based on the Rack-Cache library by Ryan Tomayko,
- * which is released under the MIT license.
- * (based on commit 02d2b48d75bcb63cf1c0c7149c077ad256542801)
```
is the rack-cache reference not true anymore? also, did you check if rack-cache refactored their thing to be more correct?

aschempp replied:
The whole class is basically a rewrite, so I think this reference is obsolete. Also, I primarily changed this because fabbot.io said so. I did not check Rack-Cache.

dbu commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -64,33 +99,116 @@ public function update(Response $response)
+    }
+
+    /**
+     * Stores the cache directive relative to the age of the response.
```
can you please explain how those things are handled? the line above essentially repeats the method name. instead, it would be good to be clear how instructions and the current age are accumulated into the age directives of the master response. (do we always end up with an age of 0, but deduct the current age of each fragment from its max-age and s-maxage?)

aschempp replied:
Good point. Would something like this make sense to you?

> Calculates the current value of max-age/s-maxage/expires by subtracting age, because the response might have been stored in the cache a while ago, so it does not expire X seconds from now, but X seconds from when it was created. It then stores the value if it is lower than any previous responses.

dbu replied:
yep. maybe like this:
```
/**
 * Lower max-age/s-maxage/expires for the whole response, if needed.
 *
 * The fragment might have been stored in cache a while ago. To keep things manageable, 
 * we substract the age so that all values are normalized for an age of 0.
 *
 * If the value is lower than the currently stored value, we update the value, to keep a rolling
 * minimal value of each instruction.
 */
```

aschempp replied:
Updated in 13febfb42effec9638d55db20434b3645f6314a2

dbu reviewed (commented):
i think its important that we improve things here. 

it is quite confusing to reason about these things, as interactions are quite complicated with the many directives. i think we need to tweak the code to be easier to understand to be more confident we get it right this time.

could you have another iteration to improve the readability of this code?

aschempp commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -64,33 +99,122 @@ public function update(Response $response)
+        $maxAge = null;
+        $sMaxage = null;
+
+        if ($this->ageDirectives['max-age']) {
```
You're right, fixed in 8fd595e84a6bb527acebbd878a9c9dadba62669d

aschempp commented on src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php:
```diff
@@ -64,33 +99,122 @@ public function update(Response $response)
+        $sMaxage = null;
+
+        if ($this->ageDirectives['max-age']) {
+            $maxAge = $this->ageDirectives['max-age'] + $this->age;
```
Well the value is that we know the age of the response, essentially when exactly the oldest "fragment" of this response was cached. Otherwise this information would be lost and every new request for the same URI would get a different response result (with updated max-age etc.).

dbu replied:
okay, i see the value of keeping that information stable.

aschempp commented:
This PR is still ready to me. Be aware that it currently points to 3.4 as a bugfix, so I'm not sure about short arrays?

I have finished everything requested at the SymfonyCon, but we haven't made any more progress since. I'm not sure who's in charge of a final decision on what needs to be completed and actually merge this?

aschempp commented:
Rebased now and updated to short arrays using the php-cs-fixer

nicolas-grekas commented:
Thanks! What about unanswered comments from David?

aschempp commented:
> Thanks! What about unanswered comments from David?

You mean this one? https://github.com/symfony/symfony/pull/26532#discussion_r242842650
I'm not sure how to handle this, or how a different comment would be an improvement…

nicolas-grekas reviewed (approved):
I'm going to trust you on this one :)

stof reviewed (approved):
This implementation is the one we discussed during the SymfonyCon hackday, solving lots of weird cases. Great to see it finally validated :smile:

fabpot commented:
Thank you @aschempp.

mpdude commented:
Great job @aschempp and all others involved!
